drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,524 @@
+#include <drivers/i2c.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tca9538, CONFIG_GPIO_LOG_LEVEL);
+
+#include "gpio_utils.h"
+
+/* TCA9538 Register addresses */
+#define TCA9538_INPUT_PORT		0x00
+#define TCA9538_OUTPUT_PORT		0x01
+#define TCA9538_CONFIGURATION		0x03
+
+/* Number of pins supported by the device */
+#define NUM_PINS 8
+
+/* Max to select all pins supported on the device. */
+#define ALL_PINS ((uint8_t)BIT_MASK(NUM_PINS))
+
+/** Cache of the output configuration and data of the pins. */
+struct tca9538_pin_state {
+	uint8_t dir;
+	uint8_t input;
+	uint8_t output;
+};
+
+struct tca9538_irq_state {
+	uint8_t rising;
+	uint8_t falling;
+};
+
+/** Runtime driver data */
+struct tca9538_drv_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	struct tca9538_pin_state pin_state;
+	struct k_sem lock;
+
+	struct gpio_callback gpio_cb;
+	struct k_work work;
+	struct tca9538_irq_state irq_state;
+	const struct device *dev;
+	/* user ISR cb */
+	sys_slist_t cb;
+};
+
+/** Configuration data */
+struct tca9538_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const struct device *i2c_dev;
+
+	bool interrupt_enabled;
+
+	const struct gpio_dt_spec gpio_int;
+
+	uint8_t i2c_addr;
+};
+
+/**
+ * @brief Gets the state of input pins of the TCA9538 I/O Port and
+ * stores in driver data struct.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int update_input(const struct device *dev)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	uint8_t cmd = TCA9538_INPUT_PORT;
+	uint8_t input_states;
+	int rc = 0;
+
+	rc = i2c_write_read(cfg->i2c_dev, cfg->i2c_addr, &cmd,
+			     sizeof(cmd), (uint8_t *)&input_states,
+			     sizeof(input_states));
+	if (rc == 0) {
+		drv_data->pin_state.input = input_states;
+	}
+	return rc;
+}
+
+/**
+ * @brief Handles interrupt triggered by the interrupt pin of TCA9538 I/O Port.
+ *
+ * If nint_gpios is configured in device tree then this will be triggered each
+ * time a gpio configured as an input changes state. The gpio input states are
+ * read in this function which clears the interrupt.
+ *
+ * @param arg Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative value for error code.
+ */
+static int gpio_tca9538_handle_interrupt(const struct device *dev)
+{
+	struct tca9538_drv_data *drv_data = dev->data;
+	struct tca9538_irq_state *irq_state = &drv_data->irq_state;
+	int rc = 0;
+	uint8_t previous_state;
+	uint8_t current_state;
+	uint8_t transitioned_pins;
+	uint8_t interrupt_mask = 0;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Store previous input state then read new value */
+	previous_state = drv_data->pin_state.input;
+	rc = update_input(dev);
```

Comment:
Since I2C access is slow we should call the function only if any of the `irq_state->rising`, `irq_state->falling` bits are enabled. I.e. if any interrupts have been configured. Even better approach would be to enable / disable the main GPIO INT via `gpio_pin_interrupt_configure_dt` in `gpio_tca9538_pin_interrupt_configure` function depending on the current requirements rather than have it always enabled in the init function.

Reply:
Exiting the function with -EINVAL if no interrupts are enabled.